### DISCUSSION TITLE
Add support for inline metadata (DataView)

### DISCUSSION THREAD

**Discussion by jeetsukumaran** [Taitava/obsidian-shellcommands#334] [Ideas]
This fantastically useful plugin now supports referencing YAML frontmatter in variable replacements, ``{{yaml_value.foo.bar}}``.

Would it be possible to add support for DataView style metadata as well? E.g., 
```
---
foo: bar
---

# Note Title

[baz:: faz]
[aaa:: bbb]

```
And being able to reference ``{{yaml_value.baz}}`` or ``{metadata.baz}}``.

**Comment by Taitava** [comment]
Hi! Thank you for your suggestion! 🙂 It's a great idea!

I haven't thought about this before, but it sounds reasonable. I need to make some desing decisions: Will it be a feature of [{{yaml_value}}](https://publish.obsidian.md/shellcommands/Variables/%7B%7Byaml_value%7D%7D) or a new variable?

## a) A new variable ⬅️ Edit 2024-07-21: I think this will be the way to go.
If it's a new variable, the name might be e.g. `{{dataview_value}}`. This clearly tells that the variable requires Dataview to be installed. (One idea in my head has been to create a variable that performs [dataview queries](https://blacksmithgu.github.io/obsidian-dataview/queries/structure/), named e.g. `{{dataview_query}}`. I haven't written about this idea anywhere, but having the part `_value` in `{{dataview_value}}` could nicely distinct the variable from the other, future coming variable).

Edit 2024-07-21: Maybe the variable could disappear from autocomplete menu, if the [Dataview](https://github.com/blacksmithgu/obsidian-dataview) plugin is not installed?

## b) Part of `{{yaml_value}}`
Pros:
- If it's part of `{{yaml_value}}`, it would be seamless to use. You wouldn't need to think about which way your data is defined, a single variable could retrieve data.

Cons:
- Harder to tell why a value cannot be found. Let's say that you are querying for `{{yaml_value:baz}}` (like in your example). If Dataview is accidentally not installed in the vault, it can't be used as a data source. The `{{yaml_value}}` variable could detect that Dataview is not present - that's ok - and so the variable will not find `baz`, as it's not present in the native YAML frontmatter. It would give an error message saying: _YAML property 'baz' is not found._ But the situation is not actually that `baz` is missing, it's just that `baz` is present, but not accessible as the Dataview plugin is not installed/enabled in this scenario. Maybe the error message could be extended to something like: _YAML property 'baz' is not found. (Note that Dataview plugin is not installed, so inline style metadata '[baz::]' could not be looked up for)_. But that addition might confuse users who do not use Dataview at all.
- The variable **name** `{{yaml_value}}` is not a 100 % perfect fit for this, as the syntax `[baz:: faz]` is not actually YAML. Maybe your other suggestion, `{{metadata}}` could suit better as a name. It could be some kind of "wrapper variable" that could look for values in both `{{yaml_value}}` and `{{dataview_value}}` (assuming I'd create that one). To streamline the naming, `{{metadata}}` could be `{{metadata_value}}`. So, three variables: `{{yaml_value}}`, `{{dataview_value}}`, and `{{metadata_value}}`. ⬅️ Edit 2024-07-21: I think this will be the way to go.

Conclusion: I'll continue thinking about this and will let you know when there's some progress 🙂. I can't promise any schedule, though.

Links for me regarding Dataview;
 - https://blacksmithgu.github.io/obsidian-dataview/resources/develop-against-dataview/
 - https://github.com/blacksmithgu/obsidian-dataview/blob/master/src/index.ts
 - https://github.com/blacksmithgu/obsidian-dataview/blob/master/src/api/plugin-api.ts

**Comment by jeetsukumaran** [comment]
Thank you for looking into this!

And yes! It makes sense to separate out ``{{yaml_value}}`` and ``{{dataview_value}}``, with a third option ``{{metadata_value}}`` as a higher level aggregator :)

**Comment by Qwerty-Space** [comment]
Hi!  I'm wondering if there's any progress on this?

---

EDIT:  It's possible to workaround this by doing something like this:
```
`= elink("obsidian://shell-commands/?vault=My%20Notes&execute=command_id&_custom_variable=" + this.thing, "Link name")`
```

**Reply by Taitava** [reply]
No progress yet.

# Anyone wanting to contribute? 🙂 

If someone is willing to use some time for this, a pull request is welcome.
- A new class could be created by copying the [`Variable_YAMLValue` class](https://github.com/Taitava/obsidian-shellcommands/blob/1785ad264c60861c30f2df15ee58bc041b3cc7e9/src/variables/Variable_YAMLValue.ts). Set `variable_name` to `dataview_value`.
- Instantiate the class in [loadVariables()](https://github.com/Taitava/obsidian-shellcommands/blob/5aeb1be2e538eb918671c6ec0eee014580d233bb/src/variables/loadVariables.ts#L71).
- Create a new helper function similar to [`getFileYAMLValue`](https://github.com/Taitava/obsidian-shellcommands/blob/1785ad264c60861c30f2df15ee58bc041b3cc7e9/src/variables/VariableHelpers.ts#L112) that reads a property from Dataview API and call that function from the variable class.
- The pull request can be a working draft - I'll do some finishing anyway.
- I'll write documentation.
- I'll create supporting variables `{{event_dataview_value}}`, `{{metadata_value}}`, and `{{event_metadata_value}}`.
- I can probably make the variable hidden in autocomplete menu if Dataview is not installed.
